artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -729,45 +733,52 @@ public void sendBuffer(ByteBuf buffer, int deliveryCount) {
    public ByteBuf getSendBuffer(int deliveryCount) {
       checkBuffer();
 
-      int amqpDeliveryCount = deliveryCount - 1;
-
-      ByteBuf result;
+      if (deliveryCount > 1) {
+         return getRedeliveredSendBuffer(deliveryCount);
 
-      if (amqpDeliveryCount > 0) {
-         // If the re-delivering the message then the header must be re-encoded
-         // (or created if not previously present).  Any delivery annotations should
-         // be skipped as well in the resulting buffer.
+      } else if (headerEnds != messagePaylodStart) {
+         return getSendBufferWithDeliveryAnnotations();
+      } else {
+         // Common case message has no delivery annotations and this is the first delivery
+         // so no re-encoding or section skipping needed.
+         return data.retainedDuplicate();
+      }
+   }
 
-         result = PooledByteBufAllocator.DEFAULT.heapBuffer(getEncodeSize());
+   private ByteBuf getSendBufferWithDeliveryAnnotations() {
```

Comment:
I might rename this to something like 'getSendBufferSkippingDeliveryAnnotations' or some such to make it clear that it is actively excluding the delivery annotations and not including them as the current name somewhat implies.